src/components/calcite-picker-item/calcite-picker-item.tsx

```diff
@@ -0,0 +1,137 @@
+
+@Component({
+  tag: "calcite-picker-item",
+  styleUrl: "./calcite-picker-item.scss",
+  shadow: true
+})
+export class CalcitePickerItem {
+  // --------------------------------------------------------------------------
+  //
+  //  Properties
+  //
+  // --------------------------------------------------------------------------
+
+  @Prop({ reflect: true }) editing = false;
+
+  @Prop() metadata: object;
+
+  @Prop({ reflect: true, mutable: true }) selected = false;
+
+  @Prop({ reflect: true }) icon: ICON_TYPES | null = null;
+
+  @Prop({ reflect: true }) textHeading: string;
+
+  @Prop({ reflect: true }) textDescription: string;
+
+  @Prop({ reflect: true }) value: string;
+
+  // --------------------------------------------------------------------------
+  //
+  //  Private Properties
+  //
+  // --------------------------------------------------------------------------
+
+  @Element() el: HTMLElement;
+
+  // --------------------------------------------------------------------------
+  //
+  //  Events
+  //
+  // --------------------------------------------------------------------------
+
+  @Event() calcitePickerItemToggled: EventEmitter;
+
+  // --------------------------------------------------------------------------
+  //
+  //  Public Methods
+  //
+  // --------------------------------------------------------------------------
+
+  @Method() async toggle(shiftPressed: boolean) {
+    this.selected = !this.selected;
+    this.calcitePickerItemToggled.emit({
+      item: this.el,
+      value: this.value,
+      selected: this.selected,
+      shiftPressed
+    });
+  }
+
+  // --------------------------------------------------------------------------
+  //
+  //  Private Methods
+  //
+  // --------------------------------------------------------------------------
+
+  iconClickHandler(e) {
+    this.toggle(e.shiftKey);
+  }
+
+  secondaryActionContainerClickHandler(e) {
+    e.stopPropagation();
+  }
+
+  // --------------------------------------------------------------------------
+  //
+  //  Render Methods
+  //
+  // --------------------------------------------------------------------------
+
+  renderIcon() {
+    const { icon } = this;
+    if (!icon) {
+      return null;
+    }
+    if (icon === ICON_TYPES.grip) {
+      return (
+        <span class="handle">
+          <CalciteIcon size="24" path={handleVertical24} />
+        </span>
+      );
+    } else {
+      const path =
+        icon === ICON_TYPES.square
+          ? this.selected
+            ? checkSquare16
+            : square16
+          : this.selected
+          ? circleFilled16F
+          : circle16F;
+      return (
+        <span class="icon">
+          <CalciteIcon size="16" path={path} />
+        </span>
+      );
+    }
+  }
+
+  render() {
+    const { icon } = this;
+    return (
+      <Host
+        class={icon !== ICON_TYPES.square && icon !== ICON_TYPES.circle ? CSS.highlight : null}
+        onClick={this.iconClickHandler.bind(this)}
```

Comment:
Bind the method in its declaration to avoid creating a new function on each render.

Reply:
Both approaches end up creating a new function on each render call. You can see this in the output code:

```js
render() {
  // ...
    h("div", {
      onClick: this.iconClickHandler.bind(this) // new fn
    },
  // ...
}
```

The way to avoid this is to bind it in the handler declaration:

```tsx
// gets bound and initialized in the constructor (after compilation)
iconClickHandler = (event: MouseEvent):void => { /* °˖✧◝(⁰▿⁰)◜✧˖° */ } 
```

which will end up something like this:

```js
 constructor() {
    this.iconClickHandler = (event) => { /* °˖✧◝(⁰▿⁰)◜✧˖° */ } 
    // ...
  }

 render() {
  // ...
    h("div", { 
      // ...
      onClick: this.iconClickHandler // handler already bound
    },
  // ...
}
```

Reply:
ah - yea didn't realize .bind created a new function. kk Changing.